feat: Support spilling for hash aggregation

kazuyukitanimura opened the pull request:
## Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/1570

## Rationale for this change

For hash aggregation, the sizes of group values and accumulators can become large that may cause out of memory.

## What changes are included in this PR?

This PR lets the hash aggregation operator spill large data to local disk using Arrow IPC format. For every input `RecordBatch`, the memory manager checks whether the new input size meets the memory configuration. If not, spilling happens, and later stream-merge sorts the spilled data to read back. The stream merge sort function was reused from the one in sort operator.


## Are these changes tested?

Yes

## Are there any user-facing changes?

No

kazuyukitanimura commented:
Waiting for https://github.com/apache/arrow-datafusion/pull/7399 for the fmt issues

alamb commented:
Thank you @kazuyukitanimura  -- I plan to review this PR tomorrow. I am very excited!

yjshen commented on datafusion/core/src/physical_plan/aggregates/row_hash.rs:
````diff
@@ -120,6 +132,56 @@ use super::AggregateExec;
+/// be re-grouped again.
+///
+/// ```text
+/// Partial Aggregation [batch_size = 2] (max memory = 3 rows)
````
For partial aggregation, when the memory is insufficient, I don't think we need to spill or merge sort; we could safely write the current hash table to output and wait for the final aggregate to do the rest.

kazuyukitanimura replied:
We could do that.
I think the trade-off is that the data size inflated by duplicated rows in terms of group values vs sorting cost. I chose to sort here because (1) the output data is compatible with non-spilling (2) chose data size benefit over sorting cost.

I would wait for other reviewers opinion on this.

alamb replied:
Other systems I have worked on have taken the approach described by @yjshen  which is to only spill the final aggregation and not spill the partial aggregates.  The rationale for not spilling the partial aggregate is that spilling is *very* expensive as it does disk IO and sorting and thus is likely orders of magnitude slower than hash partitioning more values (which is what happens if you dump less well aggregated batches).

Thus I would recommend simply not spilling at except for [`AggregateMode::Final`](https://docs.rs/datafusion/latest/datafusion/physical_plan/aggregates/enum.AggregateMode.html) or [`AggregateMode::FinalPartitioned`](https://docs.rs/datafusion/latest/datafusion/physical_plan/aggregates/enum.AggregateMode.html)

Spilling only on the Final phase might also simplify some of the accounting / schema below, as the input and the merge file input would always be the intermediate schema, I think

kazuyukitanimura replied:
Changed to spilling only on the Final phase, and output early on the partial phase.

yjshen commented on datafusion/core/src/physical_plan/aggregates/row_hash.rs:
```diff
@@ -474,7 +646,97 @@ impl GroupedHashAggregateStream {
+            let stream = read_spill_as_stream(spill, schema.clone())?;
+            streams.push(stream);
+        }
+        self.merged_stream = streaming_merge(
```
Instead of merge-sorting all spills first and then aggregating, which is a two-pass procedure, I suggest we directly merge the aggregation states by keys and produce output at the same time. 

A similar example could be found in `physical_plan/aggregates/order/full.rs`, the only difference is the number of input streams while aggregating.

kazuyukitanimura replied:
Here the merge sort is streaming, so it does not immediately merge all spills. In that way, we can take batch by batch from the top of the stream whenever the data is ready. Then we can reduce the memory pressure.
Yes I am using `GroupOrderingFull` here (couple lines below.)

yjshen replied:
Thanks, that makes sense.

yjshen reviewed (commented):
Thanks @kazuyukitanimura for working on this! This is one of the last few missing cornerstones in DataFusion. Cheers!

kazuyukitanimura commented on datafusion/core/src/physical_plan/aggregates/mod.rs:
```diff
@@ -281,6 +281,8 @@ pub struct AggregateExec {
     aggregation_ordering: Option<AggregationOrdering>,
     required_input_ordering: Option<LexOrderingReq>,
+    /// Force spilling for debugging
+    force_spill: bool,
```
Thanks, I actually tried the memory config approach. However, I was unable to reliably let spilling happen.
The reason is that `MemoryReservation.size` and the underlying `GreedyMemoryPool.used` variables are not synchronized. Thus I found it is quite difficult to spill when I want to for testing purpose. 
I may be missing something, but it seems passing a flag is more reliable.

alamb replied:
Another downside to forcing spills is that it doesn't cover the common pattern of "insert several batches until the memory limit is hit and then spill"

I think using `FairSpillPool` would likely be significantly easier to test with as it reserves a fixed portion up front for spilling operators: https://docs.rs/datafusion/latest/datafusion/execution/memory_pool/struct.FairSpillPool.html

kazuyukitanimura replied:
I figured out how to achieve this with the default `GreedyMemoryPool.used`

kazuyukitanimura commented:
Just FYI, I will be traveling entire next week. I may not be able to respond in time, but I plan to address comments the week after (9/5~)
Thank you in advance.
cc @sunchao @viirya

alamb commented on datafusion/core/src/physical_plan/aggregates/group_values/mod.rs:
```diff
@@ -42,6 +43,19 @@ pub trait GroupValues: Send {
+
+    /// Try to reserve the capacity that at least a single [`RecordBatch`] can be inserted. The
+    /// accumulator may reserve more space to speculatively avoid frequent re-allocations. After
+    /// calling try_reserve, capacity will be greater than or equal to self.len() + additional if
```
"self.len() + additional" is referenced in this comment but I don't see `additional` anywhere 🤔

kazuyukitanimura replied:
Updated

alamb commented on datafusion/physical-expr/src/aggregate/groups_accumulator/mod.rs:
```diff
@@ -157,4 +161,14 @@ pub trait GroupsAccumulator: Send {
+    fn try_reserve(&mut self, batch: &RecordBatch) -> Result<(), TryReserveError>;
+
+    /// clear the contents and shrink the capacity
+    fn clear_shrink(&mut self, batch: &RecordBatch);
```
I think `clear_shrink` may be redundant with `evaluate(EmitTo::All)`, at least with what is supposed to happen:

```
    /// ...
    /// If `emit_to`` is [`EmitTo::All`], the accumulator should
    /// return all groups and release / reset its internal state
    /// equivalent to when it was first created.
    /// ...
    fn evaluate(&mut self, emit_to: EmitTo) -> Result<ArrayRef>;
```

Maybe we can avoid adding `clear_shrink` in this case

kazuyukitanimura replied:
Removed

alamb commented on datafusion/physical-expr/src/aggregate/groups_accumulator/mod.rs:
```diff
@@ -157,4 +161,14 @@ pub trait GroupsAccumulator: Send {
     /// be `O(n)` to compute, not `O(num_groups)`
     fn size(&self) -> usize;
+
+    /// Try to reserve capacity the that at least a single [`RecordBatch`] can be inserted. The
```
While implementing the reservation up front for all accumulators results in the most accurate memory accounting, it also adds significant complexity to implementing an accumulator. This API requires each implementation to attempt to 

Another approach, that DataFusion uses elsewhere is "optimistic" memory accounting. In that case, the used memory can exceed the target slightly (~ 1 RecordBatch) in order to simplify the accounting. 

If we were to take that approach with this PR, it would look something like:

```
Insert batch into GroupsAccumulator
call GroupsAccumulator::size()
  if exceeded limit, then spill
```

And then the existing GroupsAccumulators would not need to be changed

kazuyukitanimura replied:
Applied the "optimistic" approach and removed the upfront reservations.

alamb commented on datafusion/core/src/physical_plan/aggregates/row_hash.rs:
```diff
@@ -178,6 +240,28 @@ pub(crate) struct GroupedHashAggregateStream {
 
     /// Have we seen the end of the input
     input_done: bool,
+
```
I wonder if we could encapsulate some of this state into its own structure like

```rust
struct SpillState {

    /// If data has previously been spilled, the locations of the
    /// spill files (in Arrow IPC format)
    spills: Vec<NamedTempFile>,

    /// Sorting expression for spilling batches
    spill_expr: Vec<PhysicalSortExpr>,

    /// Schema for spilling batches
    spill_schema: SchemaRef,

    /// Stream of merged outputs after spilling
    merged_stream: SendableRecordBatchStream,

    /// aggregate_arguments for merging spilled data
    merging_aggregate_arguments: Vec<Vec<Arc<dyn PhysicalExpr>>>,

    /// Force spilling always for debugging
    force_spill: bool,
}
```

kazuyukitanimura replied:
Done

alamb commented on datafusion/core/src/physical_plan/aggregates/row_hash.rs:
```diff
@@ -374,15 +519,29 @@ impl RecordBatchStream for GroupedHashAggregateStream {
         // Evaluate the grouping expressions
-        let group_by_values = evaluate_group_by(&self.group_by, &batch)?;
+        let group_by_values = if merging {
+            let group_by = PhysicalGroupBy::new_single(self.group_by.expr.clone());
```
I wonder if we need to create this PhysicalGroupBy for each batch? Why not create it once at the start?

kazuyukitanimura replied:
Updated

alamb reviewed (commented):
Thank you very much @kazuyukitanimura  -- I think this is really nicely written. I left some comments for your consideration. I think we could simplify some of the traits and memory accounting and still keep most of the benefits of this PR.

I also would love to test this myself locally, but it occurs to me I can't actually do so because datafusion CLI doesn't have any way to do it. I'll file a ticket

From my perspective the only things that would be required for merging this PR would be:
1. Performance tests showing it doesn't slow down performance (I don't expect it to and I could help with this)
2. A few more tests that exercise the spill path after aggregating more than one batch of data (I have left a suggestion on how to do that with `FairSpillPool`

Again, really nice work and thank you very much for the contribution

kazuyukitanimura commented:
Thank you @alamb for reviewing. I plan to work on addressing comments a week after next (9/5~) once I am back from my travel.

> 1. Performance tests showing it doesn't slow down performance (I don't expect it to and I could help with this)

Regarding the benchmark, what would be the best way to proceed? I read https://github.com/apache/arrow-datafusion/blob/main/benchmarks/README.md 
Is `bench.sh run tpch` a good one for this PR purpose? I am wondering if there is a github action / CI so that we can use the same machine for benchmarking...

alamb commented:
> Is bench.sh run tpch a good one for this PR purpose? I am wondering if there is a github action / CI so that we can use the same machine for benchmarking...

This is what I recommend.

I have some scripts in https://github.com/alamb/datafusion-benchmarking that I used to compare a branch to main, but they aren't super user friendly at the moment.  I'll run this branch on a gcp machine and report back

Thanks again

alamb commented:
Marking as draft to signify this PR has gotten feedback and is waiting to incorporate it before subsequent review

kazuyukitanimura commented:
Thank you @yjshen @alamb for reviewing. I addressed all comments. I would appreciate it if you could take another look. Thanks again.

alamb commented:
Thanks @kazuyukitanimura  -- I plan to review this PR tomorrow

alamb commented on datafusion/core/src/physical_plan/aggregates/row_hash.rs:
```diff
@@ -466,15 +659,119 @@ impl GroupedHashAggregateStream {
+    /// Emit the remaining rows and create a batch.
+    /// Conduct a streaming merge sort between the batch and spilled data. Since the stream is fully
+    /// sorted, set `self.group_ordering` to Full, then later we can read with [`EmitTo::First`].
+    fn update_merged_stream(&mut self) -> Result<()> {
```
At this point, rather than tracking the `merged-stream` specially, I think we could simply replace `self.input` and set the overall group by state back to `ReadingInput` to read the data from the sorted input stream (after updating `self.group_ordering` and the aggregate expressons, etc)

That might simplify the code

kazuyukitanimura replied:
Good idea. Updated.

alamb reviewed (commented):
Thank you very much @kazuyukitanimura 

I think this PR is really nicely done -- it is well documented and the code is very clean. 

There are potential future ways to simplify it I think (e.g. by reusing the existing state machines) but I think that would be fine to do as a future optimization

The only last thing I want to do prior to approving this PR is to run some benchmarks on this PR to ensure it doesn't lose performance. Otherwise it is ready to go from my perspective

kazuyukitanimura commented on datafusion/physical-expr/src/aggregate/first_last.rs:
```diff
@@ -165,6 +165,8 @@ struct FirstValueAccumulator {
     orderings: Vec<ScalarValue>,
     // Stores the applicable ordering requirement.
     ordering_req: LexOrdering,
+    // Whether merge_batch() is called before
```
Removing those fails `run_first_last_multi_partitions` with `Error: ArrowError(InvalidArgumentError("number of columns(5) must match number of fields(4) in schema"))` (at least in my local test). 

The existing code is somewhat stateful, meaning calling `state()` keeps adding a new column every time to the schema. With this PR, the final aggregation calls `merge_batch()` then, if applicable, calls `state()` for spilling via `GroupedHashAggregateStream.emit()`. The existing code is assuming that `state()` is called only once for partial aggregation. Ideally, calling `state()` multiple times should be allowed without worrying about the internal state.

viirya replied:
Hmm, why `state` call keeps adding new columns? I think it creates a new `result` and pushes values in and returns it. `result` is not share between `state` calls.

kazuyukitanimura replied:
`| a | b |` => partial aggregate => `| a | b | is_set |`  => final aggregate (spill calls state()) => `| a | b | is_set |`  => re-grouping => Done

This change avoids doubly adding `| is_set |` for spilling

viirya replied:
When `spill` is called, `emit` will be called on accumulators which calls `state` as you said. But `state`  returns a fixed vector of `ScalarValue` (e.g., for `FirstValueAccumulator` it is `[first, orderings, is_set]`). I don't see where it adds double `is_set`.

viirya replied:
I still have question about this. So marked it as unresolved.

viirya replied:
The strange part is, you only skip `is_set` in state. But actually `state` adds `first`, `orderings` too.
If your theory is `state` called twice to add doubly, `first` and `orderings` should be added doubly too.

And for other accumulators, it is same. Other accumulators might have more than one field in its state.
The strange thing is only `is_set` is skipped. And I don't see why `is_set` is so special.

kazuyukitanimura replied:
`orderings` is a vector. In reality it will be `[fisrt, orderings[0], orderings[1]..., is_set]`.
Now when final aggregation spill calls this state, `is_set` is already a part of the `orderings`

viirya replied:
I figured out the reason. This is actually a hack way to fix it. I proposed https://github.com/apache/arrow-datafusion/pull/7565 for this issue.

kazuyukitanimura commented on datafusion/core/src/physical_plan/aggregates/group_values/mod.rs:
```diff
@@ -42,6 +43,9 @@ pub trait GroupValues: Send {
     /// Emits the group values
     fn emit(&mut self, emit_to: EmitTo) -> Result<Vec<ArrayRef>>;
+
+    /// clear the contents and shrink the capacity
```
Updated

kazuyukitanimura commented on datafusion/core/src/physical_plan/aggregates/row_hash.rs:
```diff
@@ -56,6 +65,28 @@ pub(crate) enum ExecutionState {
+struct SpillState {
+    /// If data has previously been spilled, the locations of the
+    /// spill files (in Arrow IPC format)
+    spills: Vec<NamedTempFile>,
```
I will change from `NamedTempFile` to `RefCountedTempFile` once https://github.com/apache/arrow-datafusion/pull/7538/files is merged

kazuyukitanimura replied:
Updated

kazuyukitanimura commented:
Thank you @alamb for your feedbacks. I addressed them just now.
 
> The only last thing I want to do prior to approving this PR is to run some benchmarks on this PR to ensure it doesn't lose performance. Otherwise it is ready to go from my perspective

Based on the earlier conversation above, I assume that you can help on this?

alamb commented:
> Based on the earlier conversation above, I assume that you can help on this?

yes, absolutely. I am in the process of running them and will report back shortly

alamb commented:
@yjshen  would you like to review this PR again prior to merging?

Dandandan commented:
> QQuery 15

What's the reason behind QQuery 15 +2.36x faster?

yjshen commented:
On a M2 Max Mac:
```
--------------------
Benchmark tpch.json
--------------------


┏━━━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┓
┃ Query        ┃     main ┃ agg-spill-pr ┃       Change ┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━┩
│ QQuery 1     │ 373.61ms │     379.52ms │    no change │
│ QQuery 2     │  54.07ms │      71.85ms │ 1.33x slower │
│ QQuery 3     │ 127.28ms │     125.39ms │    no change │
│ QQuery 4     │  77.72ms │      77.59ms │    no change │
│ QQuery 5     │ 153.95ms │     155.71ms │    no change │
│ QQuery 6     │  74.79ms │      75.29ms │    no change │
│ QQuery 7     │ 215.22ms │     217.05ms │    no change │
│ QQuery 8     │ 181.37ms │     182.10ms │    no change │
│ QQuery 9     │ 255.84ms │     254.07ms │    no change │
│ QQuery 10    │ 201.64ms │     204.14ms │    no change │
│ QQuery 11    │  50.53ms │      51.22ms │    no change │
│ QQuery 12    │ 136.34ms │     133.59ms │    no change │
│ QQuery 13    │ 209.86ms │     213.03ms │    no change │
│ QQuery 14    │ 107.24ms │     110.68ms │    no change │
│ QQuery 15    │  77.42ms │      78.31ms │    no change │
│ QQuery 16    │  46.33ms │      46.21ms │    no change │
│ QQuery 17    │ 165.80ms │     167.61ms │    no change │
│ QQuery 18    │ 308.12ms │     299.10ms │    no change │
│ QQuery 19    │ 206.49ms │     207.36ms │    no change │
│ QQuery 20    │ 120.51ms │     122.26ms │    no change │
│ QQuery 21    │ 319.92ms │     325.50ms │    no change │
│ QQuery 22    │  47.27ms │      46.02ms │    no change │
└──────────────┴──────────┴──────────────┴──────────────┘

--------------------
Benchmark tpch_mem.json
--------------------

┏━━━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┓
┃ Query        ┃     main ┃ agg-spill-pr ┃       Change ┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━┩
│ QQuery 1     │ 249.33ms │     249.81ms │    no change │
│ QQuery 2     │  34.74ms │      43.41ms │ 1.25x slower │
│ QQuery 3     │  45.14ms │      45.13ms │    no change │
│ QQuery 4     │  35.32ms │      35.42ms │    no change │
│ QQuery 5     │  90.31ms │      89.50ms │    no change │
│ QQuery 6     │   9.60ms │       9.50ms │    no change │
│ QQuery 7     │ 172.73ms │     178.81ms │    no change │
│ QQuery 8     │  69.94ms │      68.29ms │    no change │
│ QQuery 9     │ 131.35ms │     126.52ms │    no change │
│ QQuery 10    │  84.17ms │      86.08ms │    no change │
│ QQuery 11    │  37.16ms │      36.25ms │    no change │
│ QQuery 12    │  57.98ms │      57.18ms │    no change │
│ QQuery 13    │  93.57ms │      92.33ms │    no change │
│ QQuery 14    │  11.52ms │      11.62ms │    no change │
│ QQuery 15    │  13.57ms │      14.29ms │ 1.05x slower │
│ QQuery 16    │  36.73ms │      38.23ms │    no change │
│ QQuery 17    │  74.55ms │      74.09ms │    no change │
│ QQuery 18    │ 248.72ms │     242.23ms │    no change │
│ QQuery 19    │  50.12ms │      50.65ms │    no change │
│ QQuery 20    │  73.09ms │      73.83ms │    no change │
│ QQuery 21    │ 240.83ms │     231.30ms │    no change │
│ QQuery 22    │  27.21ms │      27.18ms │    no change │
└──────────────┴──────────┴──────────────┴──────────────┘
```

sunchao reviewed (commented):
Do these benchmarks test the case of spilling?

It would be very nice if we can do some tests like in this PR: https://github.com/duckdb/duckdb/pull/7931, which measures how performance may degrade as memory budget becomes lower.

kazuyukitanimura commented on datafusion/core/src/physical_plan/aggregates/group_values/mod.rs:
```diff
@@ -42,6 +43,9 @@ pub trait GroupValues: Send {
     /// Emits the group values
     fn emit(&mut self, emit_to: EmitTo) -> Result<Vec<ArrayRef>>;
+
+    /// clear the contents and shrink the capacity to free up memory usage
```
The intention of passing `RecordBatch` is to use other information than number of rows for reserving the new capacity such as data (binary) capacity. But it will be separate future PRs.

sunchao replied:
Oh cool, perhaps its worth pointing out why `RecordBatch` is required in the comments too.

kazuyukitanimura replied:
Updated the comment

kazuyukitanimura commented:
Thank you for all the benchmarks! Glad to see there is no degradation.

> Do these benchmarks test the case of spilling?
> 
> It would be very nice if we can do some tests like in this PR: [duckdb/duckdb#7931](https://github.com/duckdb/duckdb/pull/7931), which measures how performance may degrade as memory budget becomes lower.

Thanks, though the intent of the benchmarks were to make sure this PR does not depredate, I agree that it is nice to to measure the performance under different memory pressure. Since this is the first PR for spilling, how about we do it for the future PRs?

yjshen reviewed (approved):
Thanks @kazuyukitanimura for the update! I think this PR is really neat, left one minor comment but it's none-blocking. Well done!

sunchao commented:
> Thank. Though the intent of the benchmarks were to make sure this PR does not depredate, I agree that it is nice to to measure the performance under different memory pressure. Since this is the first PR for spilling, how about we do it for the future PRs?

Sure, SGTM. Thanks @kazuyukitanimura for the work !

viirya commented on datafusion/core/src/physical_plan/aggregates/row_hash.rs:
```diff
@@ -120,6 +151,56 @@ use super::AggregateExec;
+/// The sizes of group values and accumulators can become large. Before that causes out of memory,
+/// this hash aggregator outputs those data early for partial aggregation or spills to local disk
+/// using Arrow IPC format for final aggregation. For every input [`RecordBatch`], the memory
+/// manager checks whether the new input size meets the memory configuration. If not, outputting or
+/// spilling happens. For outputting, the final aggregation takes care of re-grouping. For spilling,
```
How does it decide to output or spill?

kazuyukitanimura replied:
Looks you already got it, but just for record. A few lines above, I think this line addresses your question
```
this hash aggregator outputs those data early for partial aggregation or spills to local disk
using Arrow IPC format for final aggregation.
```

viirya commented on datafusion/core/src/physical_plan/aggregates/row_hash.rs:
```diff
@@ -120,6 +151,56 @@ use super::AggregateExec;
+/// └─────────┘    └─────────────────┘                  └─────────────────┘
+///   
+///
+/// Final Aggregation [batch_size = 2] (max memory = 3 rows)
```
Is spilling only happening during final aggregation? Seems this diagram tells this? But I don't see it is from the description. Wondering if it is true.

viirya replied:
Got the answer from reading code/doc.

viirya commented on datafusion/core/src/physical_plan/aggregates/row_hash.rs:
```diff
@@ -56,6 +65,28 @@ pub(crate) enum ExecutionState {
+    /// spill files (in Arrow IPC format)
+    spills: Vec<NamedTempFile>,
+
+    /// Sorting expression for spilling batches
+    spill_expr: Vec<PhysicalSortExpr>,
```
`sort_by_expr`?

kazuyukitanimura replied:
I would prefer to keep `spill_` prefix to be aligned with others like `spill_schema`

kazuyukitanimura commented on datafusion/core/src/physical_plan/aggregates/row_hash.rs:
```diff
@@ -318,9 +433,12 @@ impl Stream for GroupedHashAggregateStream {
                                 self.exec_state = ExecutionState::ProducingOutput(batch);
                             }
+
+                            extract_ok!(self.emit_early_if_necessary());
```
Good point, I was also thinking about this. I ended up choosing to keep the if conditions at the inner scope in order to keep the outer scope clean as much as possible. It is going to be likely I remove the `matches!(self.group_ordering, GroupOrdering::None)` from `spill_previous_if_necessary` and `emit_early_if_necessary`, and so that the changes are only in those functions. But largely TBD

viirya commented on datafusion/core/src/physical_plan/aggregates/row_hash.rs:
```diff
@@ -466,15 +625,122 @@ impl GroupedHashAggregateStream {
+        self.clear_shrink(&RecordBatch::new_empty(s));
+    }
+
+    /// Emit if the used memory exceeds the target for partial aggregation.
+    /// Currently only [`GroupOrdering::None`] is supported for spilling.
+    /// TODO: support group_ordering for spilling
+    fn emit_early_if_necessary(&mut self) -> Result<()> {
```
Is this related to spilling?

Looks like it emits early no matter spilling.

kazuyukitanimura replied:
Not related to spilling. Sorry the comment is incorrect. Updated

viirya commented on datafusion/core/src/physical_plan/aggregates/row_hash.rs:
```diff
@@ -466,15 +625,122 @@ impl GroupedHashAggregateStream {
+            && matches!(self.group_ordering, GroupOrdering::None)
+            && !matches!(self.mode, AggregateMode::Partial)
+            && !self.spill_state.is_stream_merging
+            && self.update_memory_reservation().is_err()
```
Hmm, if `update_memory_reservation` succeeds, doesn't it add additional memory reservation?

kazuyukitanimura replied:
Correct, succeeding `update_memory_reservation` means, we have more memory we can use. It lets keep accumulating.

viirya commented on datafusion/core/src/physical_plan/aggregates/row_hash.rs:
```diff
@@ -120,6 +151,56 @@ use super::AggregateExec;
+/// # Spilling
+///
+/// The sizes of group values and accumulators can become large. Before that causes out of memory,
+/// this hash aggregator outputs those data early for partial aggregation or spills to local disk
```
Using `emit` to match the API name. More easy to read accordingly with code.
```suggestion
/// this hash aggregator emits those data early for partial aggregation or spills to local disk
```

kazuyukitanimura replied:
The way I distinguish between `output` and `emit` is
`output`: output from the aggregation exec. 
`emit`: reading out from the hash table, rather more internal.
Since the subject in this sentence is `hash aggregator`, I chose `output`

viirya commented on datafusion/core/src/physical_plan/aggregates/row_hash.rs:
```diff
@@ -120,6 +151,56 @@ use super::AggregateExec;
+/// # Spilling
+///
+/// The sizes of group values and accumulators can become large. Before that causes out of memory,
+/// this hash aggregator outputs those data early for partial aggregation or spills to local disk
```
```suggestion
/// for partial aggregation, this hash aggregator emits partial states early. For final aggregation, it spills input partial state into local disk.
```

kazuyukitanimura replied:
I took some of your suggestion and revised

viirya replied:
Thank you.

viirya commented on datafusion/core/src/physical_plan/aggregates/row_hash.rs:
```diff
@@ -318,9 +433,12 @@ impl Stream for GroupedHashAggregateStream {
-                                let batch = extract_ok!(self.emit(to_emit));
+                                let batch = extract_ok!(self.emit(to_emit, false));
                                 self.exec_state = ExecutionState::ProducingOutput(batch);
                             }
+
+                            extract_ok!(self.emit_early_if_necessary());
```
```suggestion
                            } else {
                                extract_ok!(self.emit_early_if_necessary());
                            }
```

kazuyukitanimura replied:
No, `self.group_ordering.emit_to()` may return `None` for all `GroupOrdering::None`, `::Partial`, and `::Full`. So we cannot put it in the else clause...

viirya replied:
Oh right, there are `None` cases for `Parital` and `Full` too.

viirya commented on datafusion/core/src/physical_plan/aggregates/row_hash.rs:
```diff
@@ -332,8 +450,14 @@ impl Stream for GroupedHashAggregateStream {
+                            } else {
+                                // If spill files exist, stream-merge them.
+                                extract_ok!(self.update_merged_stream());
+                                self.exec_state = ExecutionState::ReadingInput;
```
Don't we need `self.input_done = false;`? I think for `ExecutionState::ReadingInput`, we assert it should be false.

viirya replied:
Nvm, saw it is set in `update_merged_stream`.

viirya commented on datafusion/core/src/physical_plan/aggregates/row_hash.rs:
```diff
@@ -466,15 +625,122 @@ impl GroupedHashAggregateStream {
+            &self.spill_state.spill_expr,
+            self.baseline_metrics.clone(),
+            self.batch_size,
+            None,
```
Shouldn't we use `GroupedHashAggregateStream.batch_size`?

kazuyukitanimura replied:
For the fetch argument, if I understand your comment correctly.
My understanding of the fetch is the number of rows that it can stop sorting and returning. Here we need all rows, so it should be `None.`

viirya replied:
Oh, yea, `fetch` is maximum output rows for the sorter. I thought it is maximum rows per batch.

viirya commented on datafusion/core/src/physical_plan/aggregates/row_hash.rs:
```diff
@@ -466,15 +625,122 @@ impl GroupedHashAggregateStream {
+    /// Emit all rows, sort them, and store them on disk.
+    fn spill(&mut self) -> Result<()> {
+        let emit = self.emit(EmitTo::All, true)?;
+        let sorted = sort_batch(&emit, &self.spill_state.spill_expr, None)?;
```
I might be missing it. But I'm wondering why we need to sort spilling batches? As they are simply partial aggregation result, and we do spilling only for `GroupOrdering::None`, does the order of spilling batches matter?

kazuyukitanimura replied:
`streaming_merge` expects all batches are pre-sorted (because it is merge sort).

viirya replied:
`streaming_merge` needs to sort because it is for sort operator. I mean for this hash aggregation, I wonder why it is needed. If it is unnecessary, that adds redundant performance cost on sorting.

kazuyukitanimura replied:
Using the example of final aggregation in the doc
https://github.com/apache/arrow-datafusion/pull/7400/files#diff-0b72cdc7a5e02b6650430c3eb7bdea4f1d2d32867b927633d92abb4fcf31c81bR187-R202
We have to re-group for rows with `a = 2`. Please notice `a = 2` can be in any spilled batches.

The spills happened because the memory is insufficient for all rows. The only choice we have left is merge based aggregation.

viirya replied:
Hmm, firstly the spilled data is partial aggregation states, right? This tries to do is to spill them and read back later after all inputs are done (all partial aggregation states are aggregated or spilled). The merged stream of spilled batches are read in as new inputs, so they are continuously aggregated into final aggregation result.

As the partial aggregation inputs should be unordered (this is not sort-based aggregation but hash aggregation), that is I wonder why we need to sort them before spilling. It doesn't matter for rows with `a = 2` in which spilled batches. Once these rows are read, they are aggregated as expected.

viirya replied:
If the sort can be removed, for now it is there simply because we want to reuse `streaming_merge`, then it is fine. But it's better to leave a TODO to remove the redundant sorting later.

kazuyukitanimura replied:
> ... so they are continuously aggregated into final aggregation result.

During this re-grouping time, we may run out of memory again. The only option is to streaming read and early outputting. In order to do so, the input for regrouping has to be sorted. Otherwise we cannot output the final aggregation results early and that makes the hash table keep all rows in memory. `a = 2` can be at very first and very last. So the very last step has to be a sort based aggregation in order to avoid OOM.

viirya replied:
It makes sense finally. This should be the main reason why sorting is necessary here.

Could you add a comment in the code for that?

kazuyukitanimura replied:
I added a short version of it to the doc section

viirya commented on datafusion/physical-expr/src/aggregate/first_last.rs:
```diff
@@ -417,7 +426,9 @@ impl Accumulator for LastValueAccumulator {
         let mut result = vec![self.last.clone()];
         result.extend(self.orderings.clone());
-        result.push(ScalarValue::Boolean(Some(self.is_set)));
+        if !self.is_merge_called {
+            result.push(ScalarValue::Boolean(Some(self.is_set)));
+        }
```
Why this is needed?

kazuyukitanimura replied:
Addressed here https://github.com/apache/arrow-datafusion/pull/7400#discussion_r1324934988

kazuyukitanimura commented:
Thank you all for the reviews! I think addressed all of them. @alamb @yjshen @sunchao @viirya 
There are some TODOs I plan to follow up in the future PRs. For any further improvements on this PR, I plan to tack on the next PR.

alamb commented:
>  Thank you all for the reviews! I think addressed all of them. @alamb @yjshen @sunchao @viirya
There are some TODOs I plan to follow up in the future PRs. For any further improvements on this PR, I plan to tack on the next PR.

Thank you @kazuyukitanimura  - this is a really nice piece of technology, as well as a nice example of collaboration. I agree let's merge this and continue working on main / follow on PRs. 

I filed https://github.com/apache/arrow-datafusion/issues/7571 to track adding spilling group benchmarks

alamb commented:
> QQuery 15

> What's the reason behind QQuery 15 +2.36x faster?


@Dandandan  I believe the reason is that Q15 is so fast (15 ms!)  it is susceptible to small pturbations, and the difference is largely a measurement error:

> │ QQuery 15    │  13.57ms │      14.29ms │ 1.05x slower │

kazuyukitanimura commented:
Great, thank you all again.

jayzhan211 commented on datafusion/core/src/physical_plan/aggregates/mod.rs:
```diff
@@ -1506,24 +1526,53 @@ mod tests {
-        ];
+        let expected = if spill {
+            vec![
+                "+---+-----+-----------------+",
```
@kazuyukitanimura Hi, do you remember why the result of `spill` is different?

eejbyfeldt replied:
I think the reason is that we are comparing the output of a partial aggregate and when `spill` we also have a lower desired batch size and hit the emit early logic: https://github.com/apache/datafusion/blob/5b6b404b9fbccb8584e8c5d2f1c366f3de736f31/datafusion/physical-plan/src/aggregates/row_hash.rs#L934-L945

viirya replied:
That's right. When spilling happens, it means we don't have enough memory to hold original batch, different partial aggregation result will be but I think it doesn't change final aggregation result.
